src/client/theme-default/Layout.vue

```diff
@@ -22,12 +22,22 @@ watch(() => route.path, closeSidebar)
 
 useCloseSidebarOnEscape(isSidebarOpen, closeSidebar)
 
-const { frontmatter } = useData()
+const { frontmatter, dir } = useData()
 
 const slots = useSlots()
 const heroImageSlotExists = computed(() => !!slots['home-hero-image'])
 
 provide('hero-image-slot-exists', heroImageSlotExists)
+
+watch(() => frontmatter.value.dir, (metaDir) => {
+  if (!inBrowser) { return }
+  const pageDir = metaDir || dir.value || 'ltr'
+  // <html dir>
+  document.documentElement.dir = pageDir
+  // <body dir> and <body style="direction">
+  document.body.dir = pageDir
+  document.body.style.direction = pageDir
+}, { immediate: true })
```

Comment:
Shouldn't this be set at build-time? Wouldn't this create Hydration mismatches? I wonder if there is a way to only set `<body dir="auto">` and let the browser choose.

Reply:
Not sure neither. I feel it's hard because these elements are out of the app. Seems [transformHtml](https://vitepress.dev/reference/site-config#transformhtml)
 is another way to help? But I feel it's too tricky.

Reply:
@posva I tried more `dir="auto"` and found it's really tricky. It only detects the direction by its first descendant element which contains text. Unless you set `dir="auto"` in every single element, the value would probably be determined by the logo label or the top ads if you set `dir="auto"` on the `<body>` which doesn't make sense.

And based on the current implementation, I also added the same logic into the build process.

Does it look good? :-)

Reply:
It should be good. I only wanted to drop the ideas I had, just in case. I can't test much at the moment. so please ignore my comments if you have already thought of it 😄